File: reflex-wasm/src/stdlib/apply.rs
// SPDX-FileCopyrightText: 2023 Marshall Wace <opensource@mwam.com>
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileContributor: Tim Kendrick <t.kendrick@mwam.com> https://github.com/timkendrickmw
use reflex::core::{uuid, ArgType, Arity, FunctionArity, Uid, Uuid};

#[derive(Default, PartialEq, Eq, Clone, Copy, Debug)]
pub struct Apply;
impl Apply {
    pub const UUID: Uuid = uuid!("faf2e936-d76c-4f54-bdae-bd67a2ab36a1");
    const ARITY: FunctionArity<2, 0> = FunctionArity {
        required: [ArgType::Strict, ArgType::Strict],
        optional: [],
        variadic: None,
    };
    pub fn arity(&self) -> Arity {
        Arity::from(&Self::ARITY)
    }
}
impl Uid for Apply {
    fn uid(&self) -> Uuid {
        Self::UUID
    }
}

File: reflex-wasm/scripts/transform/directives/apply.mjs
// SPDX-FileCopyrightText: 2023 Marshall Wace <opensource@mwam.com>
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileContributor: Tim Kendrick <t.kendrick@mwam.com> https://github.com/timkendrickmw
import { createChildScope } from '../loader.mjs';
import { NodeType } from '../parser.mjs';
import { TEMPLATE_DIRECTIVE } from './template.mjs';
import { formatSourceRange, ParseError } from '../utils.mjs';

export const APPLY_DIRECTIVE = '@apply';

export default function applyDirective(node, context) {
  const elements = node.elements.flatMap((node) =>
    context.transform ? context.transform(node, context) : [node],
  );
  const [instruction, template, ...args] = elements.filter((node) => !isNonFunctionalNode(node));
  if (
    !isNamedTermNode(APPLY_DIRECTIVE, instruction) ||
    !isNamedInstructionNode(template, TEMPLATE_DIRECTIVE) ||
    args.length !== template.elements.length - 1
  ) {
    const source = context.sources.get(context.path);
    throw new ParseError(
      node.location,
      source,
      `Invalid ${APPLY_DIRECTIVE} directive: ${formatSourceRange(source, node.location)}`,
    );
  }
  const argIdentifiers = template.elements.slice(0, -1);
  const body = template.elements[template.elements.length - 1];
  const argValues = Object.fromEntries(
    argIdentifiers.map((identifier, index) => [identifier.source, args[index]]),
  );
  const childContext = createChildScope(context, argValues);
  return context.transform ? context.transform(body, childContext) : [body];
}

function isNonFunctionalNode(node) {
  return node.type === NodeType.Whitespace || node.type == NodeType.Comment;
}

function isNamedTermNode(source, node) {
  return node.type === NodeType.Term && node.source === source;
}

function isNamedInstructionNode(node, instruction) {
  return node.type === NodeType.Instruction && node.instruction === instruction;
}
